Advanced Building Blocks - Enumerables (feature/enumerable branch)

ad9311 opened the pull request:
# Advanced Building Blocks - Enumerables
Greetings! 👋🏽
In this PR we present our Enumerables module program that contains 10 methods.
- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_any?
- my_none
- my_count
- my_map
- my_inject
- multiply_els

This program was created by @fortuneonyeka and @ad9311

Commit:
Add .rb files and linters for ruby

fortuneonyeka committed:
update enumerables.rb and test.rb

Commit:
Modify my_all? my_map methods

fortuneonyeka committed:
fix rubocop

fortuneonyeka committed:
update enumerables.rb and test.rb

Commit:
Add my_inject and multiply_els

Arrange test.rb and add missing methods

fortuneonyeka committed:
Update README.md

elisha2kyakpo1 reviewed (changes requested):
# Status: ♻️ Changes Required ♻️
**Hello guys, You have really done a good job so far, Let's just make few changes and move. Fix these methods listed below**
- [x] my_inject
- [x] my_map
- [x] my_none
- [x] my_count
- [x] my_any
- [x] my_all
- [x] my_select
- [x] my_each_with_index
- [x] my_each 
Please fix these methods because they are still giving errors.

- [x] Remove all the commented code.

Commit:
Fix my_inject methods in test.rb

The methods were not being used.

Commit:
Remove commented code

henatan99 reviewed (changes requested):
# Changes Required :recycle: 

## To Highlight 
Big or small, progress is the way to success. You made some progress and it is your first step to accomplish your task in this project. Make the necessary improvements based on the feedback given below and your project will get a lot better.

## To Improve 
Your enumerable methods have several errors and behaviors that are not the same as the standard enumerable methods. Just as samples, I have presented the checks I have run in your enumerable methods. 
array = [1, 2, 3]
1. my_each should return an Enumerator if no block is given 
    array.my_each    => your method returns a local jump error 
2. When a block is given 
   array.my_each { |i| puts i } => 1, 2
   array.each { |i| puts i } => 1, 2, 1, 2
3. my_each_with_index should return an enumerator if no block is given
    your method array.my_each_with_index  => yields local jump error 
4. my_select should return returns an enumerator if no block is given
   your method array.my_select  => raises a local jump error 
5. my_all? when no block or argument is given should return true when none of the collection members are false or nil.
   your method array.my_all?  => raises a local jump error 

There are several other errors I cannot bring them all here due to space and time constraints. Please check the ruby documentations for the enumerable methods in the link below and carefully see how the standard enumerable methods work.
[enumerables](https://ruby-doc.org/core-3.0.1/Enumerable.html) 

### Kindly make the requested changes and submit for another review

If you have any question, please contact me via:
[LinkedIn](https://www.linkedin.com/in/henok-mekonnen/) | [Twitter](https://twitter.com/henatan99) | Henok Mekonnen on Slack 

## Happy Coding :computer:

Commit:
Fix issues with methods
Improved test.rb

Commit:
Fix some errors work in progress

Commit:
Fix methods complexity

my_all? my_any? my_none?

fortuneonyeka committed:
update enumerables and test.rb rubocop implementation

DeliceLydia reviewed (changes requested):
# Status: Change Requested ![image](https://user-images.githubusercontent.com/44745855/116927431-f35ea680-ac5b-11eb-89c4-8576e166f6b3.png)

### Good job so far team you are making a progress slowly but you still have to make some changes so that your project can be approved

## To highlight

- PR is professional


### Work on the below changes and submit another CRR and don't give up 

Best,

Lydia Ingabire
Slack: Delice Lydia
[LInkedin](https://www.linkedin.com/in/delice-lydia/)

DeliceLydia commented:
### my_each

- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self

### my_each_with_index

- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self

DeliceLydia commented:
### my_all

- returns false if the block returns false or nil::range
- when no block or argument is given returns false when one of the collection members is false or nil
- when a class is passed as an argument returns false if any of the collection is not a member of such class
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

DeliceLydia commented:
### my_any

- when no block or argument is given returns false if at least one of the collection is not true
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

### my_none

- returns false if the block ever returns true for all elements::range
- when no block or argument is given returns false only if one of the collection members is true
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

DeliceLydia commented:
### my_map

- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum

fortuneonyeka committed:
update enumerable.rb

update my_each and my_each_with_index

Commit:
Fix rubocop issues

Fix my_all? my_any? my_none?

fortuneonyeka committed:
update enumerables.rb

update my_any, my_none,my_all and my_map

SajjadAhmad14 reviewed (changes requested):
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## But you have missed some of the requirements suggested by the previous reviewer 🙏

### #my_each
- [x] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] When a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

### #my_each_with_index
- [x] When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [x] When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

### #my_all
- [x] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

### Happy Coding ☺️

## You can always connect with me on slack ( @Sajjad Ahmad ).

ad9311 commented:
No problems with my_each and my_each_with_index

my_all?, my_any? and my_none? work as they should.  👇🏽

`%w[ant bear cat].my_all?(/t/)   #=> false` 

`%w[ant bear cat].my_any?(/t/)   #=> true` 

`%w[ant bear cat].my_none?(/t/)   #=> false` 

Please provided detailed information if an error is found. Also it would be helpful if visual examples are given.

george-shammar commented on enumerables.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+    arr
+  end
+
+  def my_all?(arg = nil)
+    flag = true
+    my_each do |i|
+      if block_given?
+        return flag = false unless yield(i)
+      elsif arg.instance_of?(Regexp)
+        return flag = false unless arg.match(i)
+      elsif !block_given? && arg.nil?
+        return flag = false unless i
+      else
+        return flag = false unless i.is_a?(arg)
+      end
+    end
+    flag
+  end
```
#my_all?
- [ ] When a pattern other than Regex or a Class is given, it should return true if all of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return false if any of the collection does not match the pattern.

george-shammar commented on enumerables.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+    flag
+  end
+
+  def my_any?(arg = nil)
+    flag = false
+    my_each do |i|
+      if block_given?
+        return flag = true if yield(i)
+      elsif arg.instance_of?(Regexp)
+        return flag = true if arg.match(i)
+      elsif !block_given? && arg.nil?
+        return flag = true if i
+      elsif i.is_a?(arg)
+        return flag = true
+      end
+    end
+    flag
+  end
```
#my_any

#my_any?
- [ ] When a pattern other than Regex or a Class is given, it should return false if none of the collection matches the pattern.
- [ ] When a pattern other than Regex or a Class is given, it should return true if any of the collection matches the pattern.

george-shammar commented on enumerables.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+    flag
+  end
+
+  def my_none?(arg = nil)
+    flag = true
+    my_each do |i|
+      if block_given?
+        return flag = false if yield(i)
+      elsif arg.instance_of?(Regexp)
+        return flag = false if arg.match(i)
+      elsif !block_given? && arg.nil?
+        return flag = false if i
+      elsif i.is_a?(arg)
+        return flag = false
+      end
+    end
+    flag
+  end
```
#my_none?

- [ ] When a pattern other than Regex or a Class is given, it should return false only if one of the collection matches the pattern.

george-shammar commented on enumerables.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+    count
+  end
+
+  def my_map(&arg)
+    return to_enum(:my_map) unless block_given?
+
+    arr = []
+    my_each { |i| arr << arg.call(i) }
+    arr
+  end
```
#my_map

- [ ] It should execute only the proc when both a block and a proc are given and it should return a new array with the results of running the proc once for every element in enum.

george-shammar reviewed (changes requested):
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: 

Hello , Team. :wave: 
You've made a significant progress here. And make no mistake, you've done an awesome job. Well done. :clap: :clap: :+1: :+1: 

Please, don't be fed up or get tired. There are just a little more improvements to be made to get your project approved. 

Please see my inline comments for the specific changes.

As already well articulated by your previous reviewers, your enumerable methods must pass all the test cases and perform like the regular ruby enumerables.

I have itemized the remaining test cases with brief explanation (as much as space would allow) yet to be covered by your methods.

If you feel stucked, please explore options to get help including reaching out to myself on slack.

Please submit another CRR once you're ready.

Cheers and happy coding.  :computer: :smiley:

Commit:
Add pattern condition to my_all? my_any? my_none?

Prioritize proc over block in my_map

ad9311 commented:
Fix the following issues: 

- my_all?, my_any?, my_none? all take a pattern different from Regex and Class
- my_map prioritizes the proc over the block

umairarshadbutt reviewed (changes requested):
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @ad9311 and @fortuneonyeka 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes  👇.
- [x] remove the `test.rb` file from repo.
### my_all?
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
### my_any?
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad (Code Reviewer)](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

Commit:
Fix my_all? my_any?

Commit:
Fix linter showing file already deleted

Commit:
Add test.rb to remove it properly

Commit:
Delete test.rb

Commit:
Try fix problem with linter

Linter is checking a file that does not longer exits

Commit:
Delete test.rb

Commit:
Add test.rb

Cannot delete this file as rubocop will check it in github. See previous commits for more information

ad9311 commented:
## We cannot delete test.rb file as rubocop will check it in github, linters will mark it as offence. See commits for more information.

mke2111 reviewed (changes requested):
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. 

### Well-done 💯 
    1. The methods pass most of the set tests.
    2. The readme is pretty clear.

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [x] There's one section that needs a few adjustments. Check the inline comment for this.


#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

mke2111 commented on enumerables.rb:
```diff
@@ -0,0 +1,139 @@
+
+  def my_all?(arg = nil)
+    flag = true
+    my_each do |i|
```
- [x] `my_all?` when no block or argument is given returns true when none of the collection members are false or nil.
```
Failure/Error: expect(true_array.my_all?).to be true_array.all?
     
       expected true
            got false

```
- Use this as your true_array `[1, true, 'hi', []]`. Then compare your enumerable method `my_all?` with the ruby enumerable method of `all?`. Make sure they both output the same outcomes.

Commit:
Fix my_all?

mke2111 reviewed (approved):
# 😄 🥇  STATUS: APPROVED  🥇 👏🏽 

## Hey Happy Coders, 👋🏽 
Congratulations, 🥂 . You got minimum requirements for approval. 👍🏽 

### **_Well Done_** 💯 

    1. The enumerable methods run just fine and pass all test cases. 💯 
    2. The linter checks run just fine. 👍🏽 


### You may now merge to the main branch and remember to submit a project completion form.
### I wish you the very best in your upcoming projects.

**Cheers 🍹 and Happy coding 💻** 

<hr>

### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)
